{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "56f16c78-2b3e-4042-9af8-e406c4dcba1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import time\n",
    "from Bio import motifs\n",
    "from Bio.motifs import jaspar\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "556ecb92-e563-4f55-ae26-26b88ec18984",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename columns\n",
    "def remove_rename(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df = df.drop([\"peak.1\", \"gene.1\"], axis=1)\n",
    "\n",
    "    num_cells = (len(df.columns)-4) // 2\n",
    "    new_cols = ['peak', 'gene', 'Pair', 'is_pair'] + [\"atac.\" + str(x) for x in range(num_cells)] + [\"rna.\" + str(x) for x in range(num_cells)]\n",
    "    df = df.rename(columns=dict(zip(df.columns, new_cols)))\n",
    "    return df\n",
    "\n",
    "def set_labels(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df[\"is_pair\"] = df[\"is_pair\"].apply(lambda x: 1 if x else 0)\n",
    "    return df\n",
    "    \n",
    "def normalize(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    # only counts\n",
    "    num_df = df[df.columns[4:]]\n",
    "    \n",
    "    # Summing each column (sample) to get library sizes\n",
    "    library_sizes = num_df.sum(axis=0)\n",
    "    \n",
    "    # Normalizing to CPM\n",
    "    cpm_df = num_df.div(library_sizes, axis=1) * 10**6\n",
    "    \n",
    "    # If you want to log-transform the CPM data\n",
    "    log_cpm_df = cpm_df.apply(lambda x: np.log2(x + 1))\n",
    "\n",
    "    return pd.concat([df[df.columns[:4]], log_cpm_df], axis=1)\n",
    "    \n",
    "DATA_PATH = \"../data/Tab_delimited_text/\"\n",
    "train_df = pd.read_csv(DATA_PATH + \"train.csv\")\n",
    "test_df = pd.read_csv(DATA_PATH + \"test.csv\")\n",
    "\n",
    "train_df = remove_rename(train_df)\n",
    "train_df = set_labels(train_df)\n",
    "train_df = normalize(train_df)\n",
    "\n",
    "test_df = remove_rename(test_df)\n",
    "test_df = set_labels(test_df)\n",
    "test_df = normalize(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d9c34107-08de-473d-871f-236cdde98737",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak</th>\n",
       "      <th>gene</th>\n",
       "      <th>Pair</th>\n",
       "      <th>is_pair</th>\n",
       "      <th>atac.0</th>\n",
       "      <th>atac.1</th>\n",
       "      <th>atac.2</th>\n",
       "      <th>atac.3</th>\n",
       "      <th>atac.4</th>\n",
       "      <th>atac.5</th>\n",
       "      <th>...</th>\n",
       "      <th>rna.2989</th>\n",
       "      <th>rna.2990</th>\n",
       "      <th>rna.2991</th>\n",
       "      <th>rna.2992</th>\n",
       "      <th>rna.2993</th>\n",
       "      <th>rna.2994</th>\n",
       "      <th>rna.2995</th>\n",
       "      <th>rna.2996</th>\n",
       "      <th>rna.2997</th>\n",
       "      <th>rna.2998</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1-89196985-89201657</td>\n",
       "      <td>GBP2</td>\n",
       "      <td>chr1-89196985-89201657_GBP2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>11.752078</td>\n",
       "      <td>12.693382</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.073755</td>\n",
       "      <td>11.679343</td>\n",
       "      <td>13.550867</td>\n",
       "      <td>12.522423</td>\n",
       "      <td>11.77211</td>\n",
       "      <td>13.693273</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6-33077557-33083333</td>\n",
       "      <td>HLA-DPA1</td>\n",
       "      <td>chr6-33077557-33083333_HLA-DPA1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.143825</td>\n",
       "      <td>11.839205</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.313733</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.295231</td>\n",
       "      <td>12.073922</td>\n",
       "      <td>11.679343</td>\n",
       "      <td>11.966145</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.77211</td>\n",
       "      <td>13.108365</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6-137789753-137792920</td>\n",
       "      <td>TNFAIP3</td>\n",
       "      <td>chr6-137789753-137792920_TNFAIP3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.710842</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1-212604203-212626574</td>\n",
       "      <td>ATF3</td>\n",
       "      <td>chr1-212604203-212626574_ATF3</td>\n",
       "      <td>1</td>\n",
       "      <td>14.073671</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.423905</td>\n",
       "      <td>13.170174</td>\n",
       "      <td>11.144304</td>\n",
       "      <td>15.514953</td>\n",
       "      <td>...</td>\n",
       "      <td>11.752078</td>\n",
       "      <td>12.693382</td>\n",
       "      <td>13.517459</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.679123</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.522423</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>13.108365</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2-96541661-96555628</td>\n",
       "      <td>ARID5A</td>\n",
       "      <td>chr2-96541661-96555628_ARID5A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.839008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.728735</td>\n",
       "      <td>12.707783</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.710842</td>\n",
       "      <td>12.073922</td>\n",
       "      <td>11.679343</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 6002 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       peak      gene                              Pair  \\\n",
       "0    chr1-89196985-89201657      GBP2       chr1-89196985-89201657_GBP2   \n",
       "1    chr6-33077557-33083333  HLA-DPA1   chr6-33077557-33083333_HLA-DPA1   \n",
       "2  chr6-137789753-137792920   TNFAIP3  chr6-137789753-137792920_TNFAIP3   \n",
       "3  chr1-212604203-212626574      ATF3     chr1-212604203-212626574_ATF3   \n",
       "4    chr2-96541661-96555628    ARID5A     chr2-96541661-96555628_ARID5A   \n",
       "\n",
       "   is_pair     atac.0     atac.1     atac.2     atac.3     atac.4     atac.5  \\\n",
       "0        1   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "1        1   0.000000  13.143825  11.839205   0.000000  13.313733   0.000000   \n",
       "2        1   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "3        1  14.073671   0.000000  13.423905  13.170174  11.144304  15.514953   \n",
       "4        1   0.000000   0.000000  12.839008   0.000000  13.728735  12.707783   \n",
       "\n",
       "   ...   rna.2989   rna.2990   rna.2991   rna.2992   rna.2993   rna.2994  \\\n",
       "0  ...  11.752078  12.693382   0.000000  13.073755  11.679343  13.550867   \n",
       "1  ...   0.000000   0.000000  12.295231  12.073922  11.679343  11.966145   \n",
       "2  ...   0.000000   0.000000  10.710842   0.000000   0.000000   0.000000   \n",
       "3  ...  11.752078  12.693382  13.517459   0.000000  12.679123   0.000000   \n",
       "4  ...   0.000000   0.000000  10.710842  12.073922  11.679343   0.000000   \n",
       "\n",
       "    rna.2995  rna.2996   rna.2997  rna.2998  \n",
       "0  12.522423  11.77211  13.693273       0.0  \n",
       "1   0.000000  11.77211  13.108365       0.0  \n",
       "2   0.000000   0.00000   0.000000       0.0  \n",
       "3  12.522423   0.00000  13.108365       0.0  \n",
       "4   0.000000   0.00000   0.000000       0.0  \n",
       "\n",
       "[5 rows x 6002 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "521598a5-32a1-496a-b12e-75776fa2aa36",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_correlation(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    def correl(row):\n",
    "        atac_counts = row.filter(like=\"atac.\").values.astype(float)\n",
    "        rna_counts = row.filter(like=\"rna.\").values.astype(float)\n",
    "        return np.corrcoef(atac_counts, rna_counts)[0, 1]\n",
    "\n",
    "    def covar(row):\n",
    "        atac_counts = row.filter(like=\"atac.\").values.astype(float)\n",
    "        rna_counts = row.filter(like=\"rna.\").values.astype(float)\n",
    "        return np.cov(atac_counts, rna_counts)[0, 1]\n",
    "    \n",
    "    df[\"correlation\"] = df.apply(correl, axis=1)\n",
    "    df[\"covariance\"] = df.apply(covar, axis=1)\n",
    "    return df\n",
    "\n",
    "manual_synonyms = {\n",
    "    \"C18orf25\": \"ARK2N\",\n",
    "    \"C1orf112\": \"FIRRM\",\n",
    "    \"C12orf65\": \"MTRFR\",\n",
    "    \"VNN3\": \"VNN3P\",\n",
    "}\n",
    "\n",
    "def get_gene_location(gene_name: str):\n",
    "    # apply manual synonyms\n",
    "    if gene_name in list(manual_synonyms.keys()):\n",
    "        gene_name = manual_synonyms[gene_name]\n",
    "\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    ext = f\"/lookup/symbol/homo_sapiens/{gene_name}?expand=1\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "    response = requests.get(server + ext, headers=headers)\n",
    "\n",
    "    if not response.ok:\n",
    "        print(\"Gene \" + gene_name + \" was not found in Ensembl.\")\n",
    "        return None\n",
    "\n",
    "    gene_data = response.json()\n",
    "    gene_location = {\n",
    "        \"chromosome\": gene_data[\"seq_region_name\"],\n",
    "        \"start\": int(gene_data[\"start\"]),\n",
    "        \"end\": int(gene_data[\"end\"]),\n",
    "    }\n",
    "    time.sleep(0.05)\n",
    "    return gene_location\n",
    "\n",
    "def get_peak_location(peak_str: str):\n",
    "    splits = peak_str.split(\"-\")\n",
    "    return {\n",
    "        \"chromosome\": splits[0].replace(\"chr\", \"\"),\n",
    "        \"start\": int(splits[1]),\n",
    "        \"end\": int(splits[2]),\n",
    "    }\n",
    "\n",
    "def get_peak_sequence(peak_str: str):\n",
    "    peak_loc = get_peak_location(peak_str)\n",
    "    \n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    ext = f\"/sequence/region/human/{peak_loc['chromosome']}:{peak_loc['start']}..{peak_loc['end']}:1?\"\n",
    "    headers = {\"Content-Type\": \"text/plain\"}\n",
    "\n",
    "    response = requests.get(server + ext, headers=headers)\n",
    "\n",
    "    if not response.ok:\n",
    "        print(\"Peak @ \" + peak_str + \" was not found in Ensembl.\")\n",
    "        return None\n",
    "\n",
    "    time.sleep(0.05)\n",
    "    return response.text\n",
    "\n",
    "def add_distance(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    def dist(row):\n",
    "        gene_pos = get_gene_location(row[\"gene\"])\n",
    "        peak_pos = get_peak_location(row[\"peak\"])\n",
    "        \n",
    "        if gene_pos == None:\n",
    "            return None\n",
    "            \n",
    "        if gene_pos[\"chromosome\"] != peak_pos[\"chromosome\"]:\n",
    "            print(f\"not on same chromosome: peak ({peak_pos['chromosome']}) | gene ({gene_pos['chromosome']})\")\n",
    "            return None\n",
    "\n",
    "        dist1 = abs(gene_pos[\"start\"] - peak_pos[\"end\"])\n",
    "        dist2 = abs(gene_pos[\"end\"] - peak_pos[\"start\"])\n",
    "        return min(dist1, dist2)\n",
    "\n",
    "    df[\"distance\"] = df.apply(dist, axis=1)\n",
    "    return df\n",
    "\n",
    "def add_peak_seq(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    def seq(row):\n",
    "        return get_peak_sequence(row[\"peak\"])\n",
    "\n",
    "    df[\"peak_seq\"] = df.apply(seq, axis=1)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed9b4a36-93e6-4351-9b80-f6a20a36eb61",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_df = add_correlation(train_df)\n",
    "feature_df = add_distance(feature_df)\n",
    "feature_df = add_peak_seq(feature_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "84dca6c6-2985-44a3-abb0-307e88e24626",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_df_test = add_correlation(test_df)\n",
    "feature_df_test = add_distance(feature_df_test)\n",
    "feature_df_test = add_peak_seq(feature_df_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "67f8a7de-27e9-4337-8e10-55c6bd65903b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pair</th>\n",
       "      <th>correlation</th>\n",
       "      <th>distance</th>\n",
       "      <th>peak_seq</th>\n",
       "      <th>is_pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1-89196985-89201657_GBP2</td>\n",
       "      <td>0.102802</td>\n",
       "      <td>46529</td>\n",
       "      <td>AGAATTAGAAGTTTTCTTCATTGGGAGAAGTCATGCCCTTCTTTAG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6-33077557-33083333_HLA-DPA1</td>\n",
       "      <td>0.089582</td>\n",
       "      <td>3218</td>\n",
       "      <td>AGGTGCTGGAGAGGATGTGGAGAAATAGGAACACTTTTACTCTGTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6-137789753-137792920_TNFAIP3</td>\n",
       "      <td>0.079216</td>\n",
       "      <td>74294</td>\n",
       "      <td>GCTCAAAAAAAAAAAAAAAAAGGAATAAACTACAAGTTTATGCAAC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1-212604203-212626574_ATF3</td>\n",
       "      <td>0.149961</td>\n",
       "      <td>16574</td>\n",
       "      <td>CTGAGACATCGTGTTTATGACTTGTATACTCACTCAGTTGACTGGA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2-96541661-96555628_ARID5A</td>\n",
       "      <td>0.057503</td>\n",
       "      <td>10973</td>\n",
       "      <td>ACAAGTGAGTAATGAGACAATAATCACCATCATCATCATTATTGCT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>chr1-151593645-151594727_SNX27</td>\n",
       "      <td>-0.001384</td>\n",
       "      <td>17279</td>\n",
       "      <td>TTGACAGATGAGTAAACCGAGGCCCAATGGCGGGGGTGGGGGAGTG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>chr14-52692182-52692881_PSMC6</td>\n",
       "      <td>0.025952</td>\n",
       "      <td>14297</td>\n",
       "      <td>CACCAGGCTCACTTCTTGGAAGTGTCACACCTTTAATTCATTTTTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>chr19-45527935-45529784_FOSB</td>\n",
       "      <td>0.001425</td>\n",
       "      <td>52756</td>\n",
       "      <td>GACAGGTAGTCCCCCTGGAGGAGGCCTTTGCGTGAAGGAGGGCGGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>chr17-32352719-32356100_ZNF207</td>\n",
       "      <td>0.008542</td>\n",
       "      <td>5968</td>\n",
       "      <td>ATGGCCTTGCTCTGTGGCCCAGGCTGGATTGCAGTGACATGGTTAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>chr5-77629349-77631531_WDR41</td>\n",
       "      <td>0.003987</td>\n",
       "      <td>8738</td>\n",
       "      <td>CTTCTACACTCTCACACAGTTGTGCTTAAGAAAATGAAAATAAGAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Pair  correlation  distance  \\\n",
       "0         chr1-89196985-89201657_GBP2     0.102802     46529   \n",
       "1     chr6-33077557-33083333_HLA-DPA1     0.089582      3218   \n",
       "2    chr6-137789753-137792920_TNFAIP3     0.079216     74294   \n",
       "3       chr1-212604203-212626574_ATF3     0.149961     16574   \n",
       "4       chr2-96541661-96555628_ARID5A     0.057503     10973   \n",
       "..                                ...          ...       ...   \n",
       "295    chr1-151593645-151594727_SNX27    -0.001384     17279   \n",
       "296     chr14-52692182-52692881_PSMC6     0.025952     14297   \n",
       "297      chr19-45527935-45529784_FOSB     0.001425     52756   \n",
       "298    chr17-32352719-32356100_ZNF207     0.008542      5968   \n",
       "299      chr5-77629349-77631531_WDR41     0.003987      8738   \n",
       "\n",
       "                                              peak_seq  is_pair  \n",
       "0    AGAATTAGAAGTTTTCTTCATTGGGAGAAGTCATGCCCTTCTTTAG...        1  \n",
       "1    AGGTGCTGGAGAGGATGTGGAGAAATAGGAACACTTTTACTCTGTT...        1  \n",
       "2    GCTCAAAAAAAAAAAAAAAAAGGAATAAACTACAAGTTTATGCAAC...        1  \n",
       "3    CTGAGACATCGTGTTTATGACTTGTATACTCACTCAGTTGACTGGA...        1  \n",
       "4    ACAAGTGAGTAATGAGACAATAATCACCATCATCATCATTATTGCT...        1  \n",
       "..                                                 ...      ...  \n",
       "295  TTGACAGATGAGTAAACCGAGGCCCAATGGCGGGGGTGGGGGAGTG...        0  \n",
       "296  CACCAGGCTCACTTCTTGGAAGTGTCACACCTTTAATTCATTTTTT...        0  \n",
       "297  GACAGGTAGTCCCCCTGGAGGAGGCCTTTGCGTGAAGGAGGGCGGA...        0  \n",
       "298  ATGGCCTTGCTCTGTGGCCCAGGCTGGATTGCAGTGACATGGTTAT...        0  \n",
       "299  CTTCTACACTCTCACACAGTTGTGCTTAAGAAAATGAAAATAAGAT...        0  \n",
       "\n",
       "[300 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_df[[\"Pair\", \"correlation\", \"distance\", \"peak_seq\", \"is_pair\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "11337973-13b6-4de0-ba1b-79b55c05bc0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the JASPAR PFM file\n",
    "with open(\"../data/JASPAR2024_CORE_non-redundant_pfms_jaspar.txt\") as handle:\n",
    "    records = jaspar.read(handle, \"jaspar\")\n",
    "\n",
    "motifs = []\n",
    "for motif in records:\n",
    "    pwm = motif.counts.normalize(pseudocounts=0.0001)  # Normalize with pseudocounts\n",
    "            \n",
    "    # Calculate scores\n",
    "    mat = pwm.log_odds({'A': 0.25, 'C': 0.25, 'G': 0.25, 'T': 0.25})\n",
    "\n",
    "    motifs.append(mat)\n",
    "\n",
    "def score_tf_motif(seq, in_motif_agg=\"sum\", out_motif_agg=\"mean\"):\n",
    "    sequence = Seq(seq)\n",
    "    motif_scores = []\n",
    "    for motif in motifs:\n",
    "        scores = motif.calculate(sequence)\n",
    "\n",
    "        if in_motif_agg == \"sum\":\n",
    "            motif_scores.append(np.sum(scores))\n",
    "        elif in_motif_agg == \"max\":\n",
    "            motif_scores.append(np.max(scores))\n",
    "\n",
    "    if out_motif_agg == \"mean\":\n",
    "        return np.mean(motif_scores)\n",
    "    elif out_motif_agg == \"max\":\n",
    "        return np.max(motif_scores)\n",
    "\n",
    "def raw_scores(df):\n",
    "    raw_scores = []\n",
    "    for _, row in df.iterrows():\n",
    "        sequence = Seq(row[\"peak_seq\"])\n",
    "        motif_scores = []\n",
    "        for motif in motifs:\n",
    "            scores = motif.calculate(sequence)\n",
    "            motif_scores.append(np.sum(scores))\n",
    "\n",
    "        raw_scores.append(tuple(motif_scores))\n",
    "\n",
    "    return pd.DataFrame(raw_scores, columns=[\"motif_\" + str(x) for x in range(len(motifs))])\n",
    "    \n",
    "def add_tf_scores(df, in_agg, out_agg):\n",
    "    def score(row):\n",
    "        return score_tf_motif(row[\"peak_seq\"], in_agg, out_agg)\n",
    "\n",
    "    df[\"score_\" + in_agg + \"_\" + out_agg] = df.apply(score, axis=1)\n",
    "    return df\n",
    "\n",
    "def add_raw_scores(df):\n",
    "    raw_df = raw_scores(df)\n",
    "    return pd.concat([df, raw_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cb43517e-837b-437b-a236-884a8d533493",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_df = add_tf_scores(feature_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d5212a3b-cf87-4b41-9ed8-85ca129e548f",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_df_test = add_tf_scores(feature_df_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "13e95800-e54b-4d37-9bd9-6fda7d1e31eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak</th>\n",
       "      <th>gene</th>\n",
       "      <th>Pair</th>\n",
       "      <th>is_pair</th>\n",
       "      <th>atac.0</th>\n",
       "      <th>atac.1</th>\n",
       "      <th>atac.2</th>\n",
       "      <th>atac.3</th>\n",
       "      <th>atac.4</th>\n",
       "      <th>atac.5</th>\n",
       "      <th>...</th>\n",
       "      <th>rna.2994</th>\n",
       "      <th>rna.2995</th>\n",
       "      <th>rna.2996</th>\n",
       "      <th>rna.2997</th>\n",
       "      <th>rna.2998</th>\n",
       "      <th>correlation</th>\n",
       "      <th>covariance</th>\n",
       "      <th>distance</th>\n",
       "      <th>peak_seq</th>\n",
       "      <th>peak_tf_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1-89196985-89201657</td>\n",
       "      <td>GBP2</td>\n",
       "      <td>chr1-89196985-89201657_GBP2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>13.550867</td>\n",
       "      <td>12.522423</td>\n",
       "      <td>11.77211</td>\n",
       "      <td>13.693273</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.102802</td>\n",
       "      <td>2.703429</td>\n",
       "      <td>46529</td>\n",
       "      <td>AGAATTAGAAGTTTTCTTCATTGGGAGAAGTCATGCCCTTCTTTAG...</td>\n",
       "      <td>-2.178348e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6-33077557-33083333</td>\n",
       "      <td>HLA-DPA1</td>\n",
       "      <td>chr6-33077557-33083333_HLA-DPA1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.143825</td>\n",
       "      <td>11.839205</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.313733</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>11.966145</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.77211</td>\n",
       "      <td>13.108365</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.089582</td>\n",
       "      <td>3.225886</td>\n",
       "      <td>3218</td>\n",
       "      <td>AGGTGCTGGAGAGGATGTGGAGAAATAGGAACACTTTTACTCTGTT...</td>\n",
       "      <td>-2.721785e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6-137789753-137792920</td>\n",
       "      <td>TNFAIP3</td>\n",
       "      <td>chr6-137789753-137792920_TNFAIP3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.079216</td>\n",
       "      <td>2.646761</td>\n",
       "      <td>74294</td>\n",
       "      <td>GCTCAAAAAAAAAAAAAAAAAGGAATAAACTACAAGTTTATGCAAC...</td>\n",
       "      <td>-1.476451e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1-212604203-212626574</td>\n",
       "      <td>ATF3</td>\n",
       "      <td>chr1-212604203-212626574_ATF3</td>\n",
       "      <td>1</td>\n",
       "      <td>14.073671</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.423905</td>\n",
       "      <td>13.170174</td>\n",
       "      <td>11.144304</td>\n",
       "      <td>15.514953</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.522423</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>13.108365</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.149961</td>\n",
       "      <td>4.705232</td>\n",
       "      <td>16574</td>\n",
       "      <td>CTGAGACATCGTGTTTATGACTTGTATACTCACTCAGTTGACTGGA...</td>\n",
       "      <td>-1.058393e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2-96541661-96555628</td>\n",
       "      <td>ARID5A</td>\n",
       "      <td>chr2-96541661-96555628_ARID5A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.839008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.728735</td>\n",
       "      <td>12.707783</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.057503</td>\n",
       "      <td>1.837850</td>\n",
       "      <td>10973</td>\n",
       "      <td>ACAAGTGAGTAATGAGACAATAATCACCATCATCATCATTATTGCT...</td>\n",
       "      <td>-6.657365e+05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 6007 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       peak      gene                              Pair  \\\n",
       "0    chr1-89196985-89201657      GBP2       chr1-89196985-89201657_GBP2   \n",
       "1    chr6-33077557-33083333  HLA-DPA1   chr6-33077557-33083333_HLA-DPA1   \n",
       "2  chr6-137789753-137792920   TNFAIP3  chr6-137789753-137792920_TNFAIP3   \n",
       "3  chr1-212604203-212626574      ATF3     chr1-212604203-212626574_ATF3   \n",
       "4    chr2-96541661-96555628    ARID5A     chr2-96541661-96555628_ARID5A   \n",
       "\n",
       "   is_pair     atac.0     atac.1     atac.2     atac.3     atac.4     atac.5  \\\n",
       "0        1   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "1        1   0.000000  13.143825  11.839205   0.000000  13.313733   0.000000   \n",
       "2        1   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "3        1  14.073671   0.000000  13.423905  13.170174  11.144304  15.514953   \n",
       "4        1   0.000000   0.000000  12.839008   0.000000  13.728735  12.707783   \n",
       "\n",
       "   ...   rna.2994   rna.2995  rna.2996   rna.2997  rna.2998  correlation  \\\n",
       "0  ...  13.550867  12.522423  11.77211  13.693273       0.0     0.102802   \n",
       "1  ...  11.966145   0.000000  11.77211  13.108365       0.0     0.089582   \n",
       "2  ...   0.000000   0.000000   0.00000   0.000000       0.0     0.079216   \n",
       "3  ...   0.000000  12.522423   0.00000  13.108365       0.0     0.149961   \n",
       "4  ...   0.000000   0.000000   0.00000   0.000000       0.0     0.057503   \n",
       "\n",
       "   covariance  distance                                           peak_seq  \\\n",
       "0    2.703429     46529  AGAATTAGAAGTTTTCTTCATTGGGAGAAGTCATGCCCTTCTTTAG...   \n",
       "1    3.225886      3218  AGGTGCTGGAGAGGATGTGGAGAAATAGGAACACTTTTACTCTGTT...   \n",
       "2    2.646761     74294  GCTCAAAAAAAAAAAAAAAAAGGAATAAACTACAAGTTTATGCAAC...   \n",
       "3    4.705232     16574  CTGAGACATCGTGTTTATGACTTGTATACTCACTCAGTTGACTGGA...   \n",
       "4    1.837850     10973  ACAAGTGAGTAATGAGACAATAATCACCATCATCATCATTATTGCT...   \n",
       "\n",
       "   peak_tf_score  \n",
       "0  -2.178348e+05  \n",
       "1  -2.721785e+05  \n",
       "2  -1.476451e+05  \n",
       "3  -1.058393e+06  \n",
       "4  -6.657365e+05  \n",
       "\n",
       "[5 rows x 6007 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_df.to_csv(\"../data/train_features.csv\", index=False)\n",
    "feature_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "79fef560-8cf4-4e07-8760-08d08731612e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak</th>\n",
       "      <th>gene</th>\n",
       "      <th>Pair</th>\n",
       "      <th>is_pair</th>\n",
       "      <th>atac.0</th>\n",
       "      <th>atac.1</th>\n",
       "      <th>atac.2</th>\n",
       "      <th>atac.3</th>\n",
       "      <th>atac.4</th>\n",
       "      <th>atac.5</th>\n",
       "      <th>...</th>\n",
       "      <th>rna.2994</th>\n",
       "      <th>rna.2995</th>\n",
       "      <th>rna.2996</th>\n",
       "      <th>rna.2997</th>\n",
       "      <th>rna.2998</th>\n",
       "      <th>correlation</th>\n",
       "      <th>covariance</th>\n",
       "      <th>distance</th>\n",
       "      <th>peak_seq</th>\n",
       "      <th>peak_tf_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1-1245493-1248050</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>chr1-1245493-1248050_SDF4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.727423</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003734</td>\n",
       "      <td>0.042858</td>\n",
       "      <td>13462</td>\n",
       "      <td>CGGCACCCTTTGAGGGAGGCCCGCCACCCTGCAGGGTCTCTGGAGG...</td>\n",
       "      <td>-1.236289e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1-1330394-1334148</td>\n",
       "      <td>MRPL20</td>\n",
       "      <td>chr1-1330394-1334148_MRPL20</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.324520</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.231729</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005909</td>\n",
       "      <td>0.104767</td>\n",
       "      <td>67761</td>\n",
       "      <td>TGAGTGCCCCCCAGTTCCCCTGGGAGGGCCTGCGCCTGGAGTCTGC...</td>\n",
       "      <td>-1.818039e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1-2145904-2147150</td>\n",
       "      <td>FAAP20</td>\n",
       "      <td>chr1-2145904-2147150_FAAP20</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.001309</td>\n",
       "      <td>-0.015141</td>\n",
       "      <td>37311</td>\n",
       "      <td>CTGTGACCCTGTCTGTAAAAAAAAGTTCTTAATTTTAAAAAGTCAC...</td>\n",
       "      <td>-5.889110e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1-9713011-9736481</td>\n",
       "      <td>PIK3CD</td>\n",
       "      <td>chr1-9713011-9736481_PIK3CD</td>\n",
       "      <td>1</td>\n",
       "      <td>13.812728</td>\n",
       "      <td>13.400321</td>\n",
       "      <td>14.186812</td>\n",
       "      <td>13.124375</td>\n",
       "      <td>13.304177</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.324379</td>\n",
       "      <td>14.312031</td>\n",
       "      <td>11.231729</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.054654</td>\n",
       "      <td>2.350537</td>\n",
       "      <td>16103</td>\n",
       "      <td>ATCTCTACTGAAAATACAAAAAATTAGCCAGGTGTGGTGGTAGATG...</td>\n",
       "      <td>-1.121144e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1-21287896-21301043</td>\n",
       "      <td>ECE1</td>\n",
       "      <td>chr1-21287896-21301043_ECE1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.400321</td>\n",
       "      <td>13.771800</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14.304106</td>\n",
       "      <td>12.914947</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.231429</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.067243</td>\n",
       "      <td>2.316125</td>\n",
       "      <td>57676</td>\n",
       "      <td>TCCTTTTTATACTTGACTATCTTTTTTTCTCCCCCAAGTGAGCATA...</td>\n",
       "      <td>-6.239588e+05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 6007 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     peak    gene                         Pair  is_pair  \\\n",
       "0    chr1-1245493-1248050    SDF4    chr1-1245493-1248050_SDF4        1   \n",
       "1    chr1-1330394-1334148  MRPL20  chr1-1330394-1334148_MRPL20        1   \n",
       "2    chr1-2145904-2147150  FAAP20  chr1-2145904-2147150_FAAP20        1   \n",
       "3    chr1-9713011-9736481  PIK3CD  chr1-9713011-9736481_PIK3CD        1   \n",
       "4  chr1-21287896-21301043    ECE1  chr1-21287896-21301043_ECE1        1   \n",
       "\n",
       "      atac.0     atac.1     atac.2     atac.3     atac.4     atac.5  ...  \\\n",
       "0   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000  ...   \n",
       "1   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000  ...   \n",
       "2   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000  ...   \n",
       "3  13.812728  13.400321  14.186812  13.124375  13.304177   0.000000  ...   \n",
       "4   0.000000  13.400321  13.771800   0.000000  14.304106  12.914947  ...   \n",
       "\n",
       "   rna.2994   rna.2995   rna.2996   rna.2997  rna.2998  correlation  \\\n",
       "0       0.0   0.000000  11.727423   0.000000       0.0     0.003734   \n",
       "1       0.0  12.324520   0.000000  11.231729       0.0     0.005909   \n",
       "2       0.0   0.000000   0.000000   0.000000       0.0    -0.001309   \n",
       "3       0.0  13.324379  14.312031  11.231729       0.0     0.054654   \n",
       "4       0.0   0.000000   0.000000  12.231429       0.0     0.067243   \n",
       "\n",
       "   covariance  distance                                           peak_seq  \\\n",
       "0    0.042858     13462  CGGCACCCTTTGAGGGAGGCCCGCCACCCTGCAGGGTCTCTGGAGG...   \n",
       "1    0.104767     67761  TGAGTGCCCCCCAGTTCCCCTGGGAGGGCCTGCGCCTGGAGTCTGC...   \n",
       "2   -0.015141     37311  CTGTGACCCTGTCTGTAAAAAAAAGTTCTTAATTTTAAAAAGTCAC...   \n",
       "3    2.350537     16103  ATCTCTACTGAAAATACAAAAAATTAGCCAGGTGTGGTGGTAGATG...   \n",
       "4    2.316125     57676  TCCTTTTTATACTTGACTATCTTTTTTTCTCCCCCAAGTGAGCATA...   \n",
       "\n",
       "   peak_tf_score  \n",
       "0  -1.236289e+05  \n",
       "1  -1.818039e+05  \n",
       "2  -5.889110e+04  \n",
       "3  -1.121144e+06  \n",
       "4  -6.239588e+05  \n",
       "\n",
       "[5 rows x 6007 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_df_test.to_csv(\"../data/test_features.csv\", index=False)\n",
    "feature_df_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "874fa651-2123-4d09-b0d1-d5f4dfa954b7",
   "metadata": {},
   "source": [
    "# add various tf scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "242a9f28-8b5b-46c8-a679-fd887c3d4b9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df = pd.read_csv(\"../data/train_features.csv\")\n",
    "test_df = pd.read_csv(\"../data/test_features.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9056caf5-d134-4bf4-b1a5-114618fc08fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df = train_df.drop(columns=[\"peak_tf_score\"])\n",
    "test_df = test_df.drop(columns=[\"peak_tf_score\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ba1b72c5-1101-41e2-94b1-d957526e4689",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df = add_tf_scores(train_df, \"max\", \"max\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "117d5163-e275-4aec-90ed-4471e3eef92a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    300.000000\n",
       "mean      23.300882\n",
       "std        5.852078\n",
       "min       16.136665\n",
       "25%       19.591942\n",
       "50%       21.965355\n",
       "75%       25.883785\n",
       "max       54.213387\n",
       "Name: score_max_max, dtype: float64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df.score_max_max.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "522c1a38-49ed-4534-9d63-d99d092201ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df.to_csv(\"../data/train_features_2.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b96b5d01-d79d-462a-81cb-9af2393f6665",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df = add_raw_scores(train_df)\n",
    "test_df = add_raw_scores(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4ed26a0d-0d0c-41c8-95c9-6317f3c0cb30",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "72f8fb2a-978c-4189-bf65-4944952954da",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_cols = [\"motif_\" + str(x) for x in range(len(motifs))]\n",
    "def normalize_motif_cols(df):\n",
    "    scaler = StandardScaler()\n",
    "    for col in motif_cols:\n",
    "        df[col] = scaler.fit_transform(df[[col]])\n",
    "\n",
    "    return df\n",
    "\n",
    "def add_pca_components(df, n_components):\n",
    "    pca = PCA(n_components=n_components)\n",
    "    pca_components = pca.fit_transform(df[motif_cols])\n",
    "\n",
    "    # Create new column names for the principal components\n",
    "    component_names = [f'pc_{i+1}' for i in range(n_components)]\n",
    "    \n",
    "    # Create a DataFrame with the PCA components\n",
    "    pca_df = pd.DataFrame(data=pca_components, columns=component_names)\n",
    "    \n",
    "    # Concatenate the original DataFrame with the PCA components\n",
    "    df_with_pca = pd.concat([df, pca_df], axis=1)\n",
    "    \n",
    "    return df_with_pca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cc493705-df19-4df1-a970-53b01331a367",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_pca_df = normalize_motif_cols(train_df)\n",
    "train_pca_df = add_pca_components(train_pca_df, n_components=20)\n",
    "\n",
    "test_pca_df = normalize_motif_cols(test_df)\n",
    "test_pca_df = add_pca_components(test_pca_df, n_components=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "861fe88d-206f-4999-870b-6c2b22d791eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_pca_df.to_csv(\"../data/train_features_3.csv\", index=False)\n",
    "test_pca_df.to_csv(\"../data/test_features_3.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e6a3e77-6336-48e8-b96f-50d1ac153c11",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
